Support encrypting binary columns

djmb opened the pull request:
### Motivation / Background

ActiveRecord Encryption doesn't prevent you from encrypting binary columns but it doesn't have proper support for it either.

When the data is fed through encrypt/decrypt it is converted to a String. This means that the the encryption layer is not transparent to binary data - which should be passed as Type::Binary::Data.

As a result the data is not properly escaped in the SQL queries or deserialized correctly after decryption.

However it just happens to work fine for MySQL and SQLite because the MessageSerializer doesn't use any characters that need to be encoded. However if you try to use a custom serializer that does then it breaks.

PostgreSQL on the other hand does not work - because the Bytea type is passed a String rather than a Type::Binary::Data to deserialize, it attempts to unescape the data and either mangles it or raises an error if it contains null bytes.

### Detail

The commit fixes the issue, by reserializing the data after encryption and decryption. For text data that's a no-op, but for binary data we'll convert it back to a Type::Binary::Data.

djmb committed:
Support encrypting binary columns

ActiveRecord Encryption doesn't prevent you from encrypting binary
columns but it doesn't have proper support for it either.

When the data is fed through encrypt/decrypt it is converted to a
String. This means that the the encryption layer is not transparent
to binary data - which should be passed as Type::Binary::Data.

As a result the data is not properly escaped in the SQL queries or
deserialized correctly after decryption.

However it just happens to work fine for MySQL and SQLite because the
MessageSerializer doesn't use any characters that need to be encoded.
However if you try to use a custom serializer that does then it breaks.

PostgreSQL on the other hand does not work - because the Bytea type
is passed a String rather than a Type::Binary::Data to deserialize, it
attempts to unescape the data and either mangles it or raises an error
if it contains null bytes.

The commit fixes the issue, by reserializing the data after
encryption and decryption. For text data that's a no-op, but for binary
data we'll convert it back to a Type::Binary::Data.

djmb commented on activerecord/lib/active_record/encryption/encrypted_attribute_type.rb:
```diff
@@ -150,6 +152,14 @@ def decryption_options
           @clean_text_scheme ||= ActiveRecord::Encryption::Scheme.new(downcase: downcase?, encryptor: ActiveRecord::Encryption::NullEncryptor.new)
         end
+
+        def reserialize(value)
```
Rather than doing this here, we could have an `EncryptedBinaryAttributeType` that subclasses ::ActiveModel::Type::Binary` instead and extract the common behaviour to a module.

jorgemanrubia commented on activerecord/lib/active_record/encryption/encrypted_attribute_type.rb:
```diff
@@ -85,11 +85,13 @@ def decrypt(value)
               end
             end
+
+            reserialize(decrypted_value)
```
I'd consider extracting the `unless` statement to its own `#decrypt_as_text(value)` method, and then do:

```ruby
reserialize decrypt_as_text(value) 
```

And, for symmetry, do the same with a `#encrypt_as_text(value)` in #44052 

```ruby
deserialize encrypt_as_text(value)
```

djmb replied:
I've updated that here - https://github.com/rails/rails/pull/50920/commits/16330f172a9b0ec0396832a58674b1f4d2a2cb50 👍

jorgemanrubia reviewed (approved):
Great one @djmb. Left a minor suggestion for your consideration.

I'd also link this PR with the PostgreSQL issue. Will this fix it out of the box?

djmb committed:
Extract decrypt_as_text/encrypt_as_text

djmb commented:
Thanks @jorgemanrubia!

There isn't a Rails issue for this - one was raised on the SolidCache repo though (https://github.com/rails/solid_cache/issues/102)

jeremy commented on activerecord/lib/active_record/encryption/encrypted_attribute_type.rb:
```diff
@@ -150,6 +158,14 @@ def decryption_options
         end
+
+        def reserialize(value)
+          if cast_type.binary?
```
Is this the lone case needing reserialization?

djmb replied:
Text and binary columns are the only ones able to store the encrypted JSON blobs.

I checked whether we could always serialize here instead, in the hope that it would be a no-op for String columns, but type serialization is much more complicated than that - it might be converting serialized or store values for instance.

I've changed this in https://github.com/rails/rails/pull/50920/commits/38b7df7aa2629e27c1afa7a0d2eb2c961077a1aa, now we just wrap strings as Binary::Data, as that's actually reversing what the encryptor does. I've also renamed the method as "serialize" is overloaded with meanings here.

djmb committed:
Handle serialized binary data in encrypted columns

Calling `serialize` is not always possible, because the column type
might not expect to be serializing a String, for example when declared
as serialzed or store attribute.

With binary data the encryptor was passed an
`ActiveModel::Type::Binary::Data`` and returned a `String``. In order to
remain transparent we need to turn the data back into a
`ActiveModel::Type::Binary::Data` before passing it on.

We'll also rename `serialize`` to `text_to_database_type` to be a bit
more descriptive.
